src/status_im/contexts/wallet/wallet_connect/events.cljs

```diff
@@ -204,10 +198,59 @@
                                          {:version version}))}]]]}
        {:fx [[:dispatch [:wallet-connect/pair scanned-text]]]}))))
 
+;; We first load sessions from database, then we initiate a call to Wallet Connect SDK and
+;; then replace the list we have stored in the database with the one that came from the SDK.
+;; In addition to that, we also update the backend state by marking sessions that are not
+;; active anymore by calling `:wallet-connect/disconnect-session`.
+(rf/reg-event-fx
+ :wallet-connect/fetch-active-sessions-success
+ (fn [{:keys [db]} [sessions]]
+   (let [time-now           (-> (time/now)
+                                time.coerce/to-long
+                                (/ 1000))
+         persisted-sessions (:wallet-connect/sessions db)
+         sessions           (->> sessions
+                                 vals
+                                 (map wallet-connect-core/sdk-session->db-session))
+         expired-sessions   (remove
+                             (fn [{:keys [expiry]}]
+                               (> expiry time-now))
+                             persisted-sessions)]
+     {:fx (mapv (fn [{:keys [pairingTopic]}]
+                  [:wallet-connect/disconnect-session pairingTopic])
+                expired-sessions)
+      :db (assoc db :wallet-connect/sessions sessions)})))
+
+(rf/reg-event-fx
+ :wallet-connect/fetch-active-sessions
+ (fn [{:keys [db]}]
+   (let [web3-wallet (get db :wallet-connect/web3-wallet)]
+     {:fx [[:effects.wallet-connect/fetch-active-sessions
+            {:web3-wallet web3-wallet
+             :on-fail     #(log/error "Failed to get active sessions" {:error %})
+             :on-success  #(rf/dispatch [:wallet-connect/fetch-active-sessions-success
+                                         (js->clj % :keywordize-keys true)])}]]})))
+
 (rf/reg-event-fx
  :wallet-connect/fetch-persisted-sessions-success
  (fn [{:keys [db]} [sessions]]
-   {:db (assoc db :wallet-connect/persisted-sessions sessions)}))
+   (let [sessions' (mapv (fn [{:keys [sessionJson] :as session}]
+                           (assoc session
+                                  :accounts
+                                  (-> sessionJson
+                                      types/json->clj
+                                      :namespaces
+                                      :eip155
+                                      :accounts)))
+                         sessions)]
+     {:fx [[:dispatch [:wallet-connect/fetch-active-sessions]]]
```

Comment:
Maybe we should put all this logic of figuring out the sessions based on the active and persisted ones inside one effect (or better a function used by the effect). It could be in its own namespace i.e. `sessions`. Otherwise it might become difficult to manage & read with all the different events and "success/failed" events, which call other effects and so on. It's not necessary to change it now as it might be late for that with the deadline and all, but IMO would be really good to move all this logic outside the events eventually and try to keep the logic in functions instead of directly inside events/effects that call each other. What do you think?

Reply:
Let me create an issue and handle it separately if you don't mind. Gonna be a bit of a nice refactoring but maybe better to handle it after we finish with the core functionality.